docs/spec.md

```diff
@@ -140,11 +140,29 @@ which may allow us to remove tuple types from StableHLO
 
 ```ebnf
 ElementType ::= BooleanType | IntegerType | FloatType | ComplexType
+              | QuantizedType
 BooleanType ::= 'i1'
 IntegerType ::= 'si4' | 'si8' | 'si16' | 'si32' | 'si64'
               | 'ui4' | 'ui8' | 'ui16' | 'ui32' | 'ui64'
 FloatType   ::= 'f8E4M3FN' | 'f8E5M2' | 'bf16' | 'f16' | 'f32' | 'f64'
 ComplexType ::= 'complex' '<' ('f32' | 'f64') '>'
+
+QuantizedType ::= '!quant.uniform' '<' QuantizationStorageTypeSpec
+                ':' QuantizationExpressedType
+                [':' QuantizationDimension]
+                ',' QuantizationScalesAndZeroPoints '>'
+QuantizationStorageTypeSpec ::= QuantizationStorageType
+                              ['<' QuantizationStorageMin ':' QuantizationStorageMax '>']
+QuantizationStorageType ::= IntegerType
+QuantizationStorageMin ::= IntegerConstant
+QuantizationStorageMax ::= IntegerConstant
+QuantizationExpressedType ::= FloatType
+QuantizationDimension ::= IntegerConstant
+QuantizationScalesAndZeroPoints ::= (QuantizationScaleAndZero
+                                  | '{' QuantizationScaleAndZero {',' QuantizationScaleAndZero} '}')
+QuantizationScaleAndZero ::= QuantizationScale ':' QuantizationZeroPoint
+QuantizationScale ::=  FloatConstant
```

Comment:
@dominicsymes Following up on integer only processing, what if we had two different versions of QuantizedType - one as proposed in the PR, and another one where a floating-point scale is replaced with integer multiplier and shift. Would that work for you with respect to #1149?

This idea was proposed by @sngyhan in an internal discussion, and I think that it's pretty neat. I think that it allows implementations to more easily detect different styles of quantization, which is nice. Also, through FP8, we already have a precedent for providing multiple ways of doing similar things through similar types, and this proposal goes along the same lines. Please let us know what you think!

Reply:
@dominicsymes is out of the office at the moment. I won't speak for him, but I'll add a bit here, and he can add more when he's back.
In my view, it's not a completely bad option. If we implemented this, there should be a pass that was able to convert between the two formats with a fully defined result. If there is agreement on that pass, then all integer based implementations would agree, and an integer based quantized tensor could be converted to a floating-point version.

Reply:
Hi. Maybe in future we can extend QuantizationScale to accept a type for multiplier/shift to support the case with the defined conversion format between the two.

Reply:
Hi @burmako. @sngyhan

Thanks for your comments. Having a quantization type with integer multiplier and shift does get part of the way there provided (as @eric-k256  mentions) that there is a defined conversion between the two formats.
For implementations to agree it also needs to be defined how to convert these into actual scales to be applied.
In an element-wise multiply operation the scale applied after the multiply is the ratio of the product of the input scales to the output scale. Similarly, in the element-wise add in this PR it needs be defined how the integer scales apply as input and output scales.
An advantage of separating out these scale changes is that they can then be handled in a standard way rather than being defined as part of each operation.

Reply:
Hi @Maratyszcza 
Thanks for the comments! 

We already have (1). I have added (2)
```
* (C6) For all `i`, `scales[i] > 0`.
* (C7) For all `i`, `is_finite(scales[i]) = true`.
```

>> About (3): .... assures compatibility with systems which don't support subnormal numbers where they will treated as 0.

Maybe I did not get this. 
Isn't the constraint will be restrictive to the systems supporting sub-normals? 
If this check is relevant when converting a double multiplier (e.g. `S1*S2/S3` in eq (5) in https://arxiv.org/pdf/1712.05877.pdf to a int multiplier/shift), then does it matter if we add constraints to the individual scales? That is, isn't the double multiplier can still be sub-normal is the individual scales are not. 

On a side note: For (3), Can we explore this constraint when we deal with the "integer based quantized type" in a separate PR (as suggested in https://github.com/openxla/stablehlo/pull/1352#pullrequestreview-1378179457) ?

Reply:
+1 for opening the possibility of both scaling mechanisms.

Reply:
As discussed above, let's move the work on integer-based QuantizedTensorType to a separate PR, given that there's broad interest in exploring this. We opened #1404 to track this.